### PR TITLE
 Feat/add access

### PR DESCRIPTION
**Pull request by matiu**
- removes 'isTemporaryPubKey'
- adds support to 'addAccess' endpoint at BWS
- refactor legacyImport, to use addAccess.
- adds mnemonics support (BIP39)

### CODE REVIEW

**Comment by matiu** [comment]
Updated, ready for review. needs https://github.com/bitpay/bitcore-wallet-service/pull/311

**Comment by Kirvx** [comment]
Please don't forget the french wordlist :)
https://github.com/bitcoin/bips/blob/master/bip-0039/french.txt

**Comment by matiu** [comment]
hey @Kirvx, could you send a PR to https://github.com/bitpay/bitcore-mnemonic with the FR wordlist? the words are at https://github.com/bitpay/bitcore-mnemonic/tree/master/lib/words

**Comment by Kirvx** [comment]
@matiu Sorry I can't, I'm not a dev :/

**Comment by greggzigler** [comment]
@kirvx, see https://github.com/bitpay/bitcore-mnemonic/pull/28 for french mnemonic wordlist

**Comment by matiu** [comment]
GREAT!! :)

On Mon, Aug 10, 2015 at 3:01 PM, Gregg Zigler <notifications@github.com>
wrote:

> @kirvx <https://github.com/kirvx>, see bitpay/bitcore-mnemonic#28
> <https://github.com/bitpay/bitcore-mnemonic/pull/28> for french mnemonic
> wordlist
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bitpay/bitcore-wallet-client/pull/107#issuecomment-129550920>
> .
>



-- 
Matías Alejo Garcia
@ematiu
Roads? Where we're going, we don't need roads!

**Comment by Kirvx** [comment]
Thanks :smiley:

**Comment by matiu** [comment]
updated, ready for review.

**Comment by isocolsky** [comment]
needs rebase

**Comment by matiu** [comment]
updated.

File: lib/api.js [hunk 1]
```diff
@@ -288,21 +324,131 @@ API.prototype.export = function(opts) {
...
   }
-  this.credentials = credentials;
 };
 
+
+API.prototype.importFromMnemonic = function(words, opts, cb) {
+  log.info('Importing from 12 Words');
+
+  var nrWords = words.split(' ').map(function(v){
+    return _.trim(v);
```
Comment (isocolsky) [inline lib/api.js]: I don't think the `map` function is needed here

File: lib/api.js [hunk 2]
```diff
@@ -288,21 +324,131 @@ API.prototype.export = function(opts) {
...
+  }).length;
+  if (nrWords != 12)
+    return cb(Errors.INVALID_BACKUP);
+
+  var key = Credentials.keyFromMnemonic(words, opts.passphrase, opts.network);
```
Comment (isocolsky) [inline lib/api.js]: Since the parent function is async, any exception thrown here should be returned using the callback

**Comment by matiu** [comment]
updated.

File: lib/api.js [hunk 3]
```diff
@@ -288,21 +325,133 @@ API.prototype.export = function(opts) {
...
+      if (!self.credentials.network || !opts.m || !opts.n)
+        return cb(Errors.MISSING_ARGS_TO_RECOVER_WALLET);
+
+      if (opts.n > 1 && !opts.publicKeyRing)
+        return cb(Errors.MISSING_ARGS_TO_RECOVER_WALLET);
+
+      self.credentials.m = opts.m;
+      self.credentials.n = opts.n;
+
+      // TODO: n>1 ?
```
Comment (isocolsky) [inline lib/api.js]: What does this mean?

File: lib/api.js [hunk 4]
```diff
@@ -288,21 +325,115 @@ API.prototype.export = function(opts) {
 API.prototype.import = function(str, opts) {
   opts = opts || {};
 
-  var credentials;
+
+  // Json input?
```
Comment (isocolsky) [inline lib/api.js]: This comment is missplaced and I think not really needed?

File: lib/api.js [hunk 5]
```diff
@@ -288,21 +325,115 @@ API.prototype.export = function(opts) {
...
+API.prototype.importFromExtendedPrivateKey = function(xPrivKey, cb) {
+  var self = this;
+
+  log.debug('Importing from Extended Private Key');
+  this.credentials = Credentials.fromExtendedPrivateKey(xPrivKey);
+
+  // First option, grab wallet info from BWS.
+  self.openWallet(function(err, ret) {
+    // it worked?
+    // TODO download encrypted Wallet key ...
```
Comment (isocolsky) [inline lib/api.js]: not TODO anymore...

File: lib/errors/errordefinitions.js [hunk 6]
```diff
@@ -0,0 +1,23 @@
+'use strict';
+
+var _ = require('lodash');
+
+var ClientError = require('./clienterror');
+
+var errors = {
+  INVALID_BACKUP: 'Invalid Backup',
+  WALLET_DOES_NOT_EXISTS: 'Wallet does not exists. Need to recreate',
```
Comment (isocolsky) [inline lib/errors/errordefinitions.js]: small typo: should be `WALLET_DOES_NOT_EXIST`

File: lib/api.js [hunk 7]
```diff
@@ -288,21 +325,115 @@ API.prototype.export = function(opts) {
...
+    if (ret) return cb(null, ret);
+
+    log.info('Copayer not found, trying to add access');
+
+    //Second option, lets try to add an access
+    self.addAccess({}, function(err) {
+
+      // it worked?
+      if (!err) 
+        self.openWallet(cb);
```
Comment (isocolsky) [inline lib/api.js]: Isn't this too optimistic? What if the wallet cannot be opened for some other reason? Won't this cause the app to fill up the access key list?

**Comment by matiu** [comment]
updated.

File: lib/api.js [hunk 8]
```diff
@@ -1442,43 +1548,45 @@ API.prototype.createWalletFromOldCopay = function(username, password, blob, cb)
...
+    signature: sig,
+    name: opts.name,
+    restrictions: opts.restrictions,
+  };
 
-  this._doPutRequest('/v1/copayers/', args, function(err, wallet) {
+  this._doPutRequest('/v1/copayers/' + copayerId + '/', opts, function(err, res) {
     if (err) return cb(err);
-    return cb(null, wallet);
+    return cb(null, res.wallet, reqPrivKey);
```
Comment (isocolsky) [inline lib/api.js]: It would be cleaner to return a single result object IMO.

File: lib/credentials.js [hunk 9]
```diff
@@ -60,9 +91,38 @@ Credentials.fromExtendedPrivateKey = function(xPrivKey) {
   return x;
 };
 
-Credentials.fromExternalWalletPublicKey = function(xPubKey, source, index) {
+
+Credentials.keyFromMnemonic = function(words, passphrase, network) {
```
Comment (isocolsky) [inline lib/credentials.js]: This looks like an internal method, I think prepending a `_` would make the API cleaner.

**Comment by matiu** [comment]
updated

**Comment by isocolsky** [comment]
Amazing job! Thanks!

**Pull request merged**